libs/x-content/src/main/java/org/elasticsearch/xcontent/Text.java

```diff
@@ -72,11 +81,21 @@ public boolean hasString() {
         return text != null;
     }
 
-    /**
-     * Returns a {@link String} view of the data.
-     */
+    @Override
     public String string() {
-        return text == null ? bytes.utf8ToString() : text;
+        if (text == null) {
+            var byteBuff = ByteBuffer.wrap(bytes.bytes(), bytes.offset(), bytes.length());
+            text = StandardCharsets.UTF_8.decode(byteBuff).toString();
```

Comment:
Re my comment above: I think we could follow this with:

```
assert (stringLength < 0) || (stringLength == text.length());
```

This would catch cases where the wrong string length is specified at construction time.